tests/filtered_relation/tests.py

```diff
@@ -0,0 +1,351 @@
+                                             condition=Q(book__title__icontains='book',
+                                                         book__editor_id=self.editor_a.pk)
+                                             ).filter(book_editor_a__isnull=False),
+            ["<Author: Alice>"])
+
+    def test_filtered_relation_multiple_times(self):
+        self.assertQuerysetEqual(
+            Author.objects.filtered_relation('book', alias='book_title_alice',
+                                             condition=Q(book__title__icontains='alice')
+                                             ).filter(book_title_alice__isnull=False).filter(
+                                                 book_title_alice__isnull=False).distinct(),
+            ["<Author: Alice>"])
+
+    def test_exclude_relation_with_join(self):
+        self.assertQuerysetEqual(
+            Author.objects.filtered_relation(
+                'book', alias='book_alice',
+                condition=~Q(book__title__icontains='alice')
+            ).filter(book_alice__isnull=False).distinct(),
+            ["<Author: Jane>"])
+
+    def test_filtered_relation_with_m2m(self):
+        qs = Author.objects.filtered_relation(
+            'favourite_books', alias='favourite_books_written_by_jane',
+            condition=Q(favourite_books__in=[self.book2])).filter(favourite_books_written_by_jane__isnull=False)
+        self.assertQuerysetEqual(qs, ["<Author: Alice>"])
+
+    def test_filtered_relation_with_m2m_deep(self):
+        qs = Author.objects.filtered_relation(
+            'favourite_books', alias='favourite_books_written_by_jane',
+            condition=Q(favourite_books__author=self.author2)
+        ).filter(favourite_books_written_by_jane__title='The book by Jane B')
+        self.assertQuerysetEqual(qs, ["<Author: Alice>"])
+
+    def test_filtered_relation_with_m2m_multijoin(self):
+        qs = Author.objects.filtered_relation(
+            'favourite_books', alias='favourite_books_written_by_jane',
+            condition=Q(favourite_books__author=self.author2)
+        ).filter(favourite_books_written_by_jane__editor__name='b').distinct()
+        self.assertQuerysetEqual(qs, ["<Author: Alice>"])
+
+    def test_filtered_relation_values_list(self):
+        self.assertQuerysetEqual(Author.objects.filtered_relation(
+            'book', alias='book_alice',
+            condition=Q(book__title__iexact='poem by alice')
+        ).values_list('book_alice__title', flat=True).order_by(F('book_alice__title').desc(nulls_last=True)),
+            ["Poem by Alice", None], lambda x: x)
+
+    def test_filtered_relation_values(self):
+        self.assertQuerysetEqual(Author.objects.filtered_relation(
+            'book', alias='book_alice',
+            condition=Q(book__title__iexact='poem by alice')).filter(book_alice__isnull=False).values(),
+            [{'id': 1, 'name': 'Alice'}], lambda x: x)
```

Comment:
Is `lambda x: x` really needed?

Reply:
Yes, otherwise we end up comparing a string representation of the dictionary. Since ordering is not guaranteed during type casting, it is simpler to compare plain dict objects.

Reply:
It looks a bit like https://code.djangoproject.com/ticket/27203 I could be wrong, though.

Reply:
Nice, it does the trick.